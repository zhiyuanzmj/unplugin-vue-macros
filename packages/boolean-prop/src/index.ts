import {
  type ConstantTypes,
  type NodeTransform,
  type NodeTypes,
} from '@vue/compiler-core'

export function transformBooleanProp(): NodeTransform {
  return (node) => {
    if (node.type !== (1 satisfies NodeTypes.ELEMENT)) return
    for (const [i, prop] of node.props.entries()) {
      if (
        prop.type !== (6 satisfies NodeTypes.ATTRIBUTE) ||
        prop.value !== undefined
      )
        continue
      node.props[i] = {
        type: 7 satisfies NodeTypes.DIRECTIVE,
        name: 'bind',
        arg: {
          type: 4 satisfies NodeTypes.SIMPLE_EXPRESSION,
          constType: 3 satisfies ConstantTypes.CAN_STRINGIFY,
          content: 'checked',
          isStatic: true,
          loc: prop.loc,
        },
        exp: {
          type: 4 satisfies NodeTypes.SIMPLE_EXPRESSION,
          constType: 3 satisfies ConstantTypes.CAN_STRINGIFY,
          content: 'true',
          isStatic: false,
          loc: prop.loc,
        },
        loc: prop.loc,
        modifiers: [],
      }
    }
  }
}
