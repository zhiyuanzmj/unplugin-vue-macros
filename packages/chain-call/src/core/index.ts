import {
  DEFINE_PROPS,
  MagicString,
  WITH_DEFAULTS,
  generateTransform,
  isCallOf,
  isIdentifierOf,
  parseSFC,
  removeMacroImport,
  walkAST,
} from '@vue-macros/common'
import {
  type CallExpression,
  type MemberExpression,
  type Node,
} from '@babel/types'

export function transformChainCall(code: string, id: string) {
  if (!code.includes(DEFINE_PROPS)) return

  const { scriptSetup, getSetupAst, offset } = parseSFC(code, id)
  if (!scriptSetup) return

  const s = new MagicString(code)
  const setupAst = getSetupAst()!

  walkAST<Node>(setupAst, {
    enter(node) {
      if (removeMacroImport(node, s, offset)) return
      if (isChainCall(node)) processChainCall(node)
    },
  })

  function processChainCall(node: CallExpression) {
    const withDefaultString: string | undefined =
      node.arguments[0] && s.sliceNode(node.arguments[0], { offset })
    const definePropsString = s.sliceNode(
      (node.callee as MemberExpression).object as CallExpression,
      { offset }
    )

    s.overwriteNode(
      node,
      withDefaultString
        ? `${WITH_DEFAULTS}(${definePropsString}, ${withDefaultString})`
        : definePropsString,
      { offset }
    )
  }

  return generateTransform(s, id)
}

function isChainCall(node: Node): node is CallExpression {
  // defineProps().withDefaults()
  return (
    node.type === 'CallExpression' &&
    node.callee.type === 'MemberExpression' &&
    isCallOf(node.callee.object, DEFINE_PROPS) &&
    isIdentifierOf(node.callee.property, WITH_DEFAULTS)
  )
}
