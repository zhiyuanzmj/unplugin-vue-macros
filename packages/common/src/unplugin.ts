import {
  type FilterPattern,
  createFilter as createRollupFilter,
} from '@rollup/pluginutils'
import { generateTransform } from 'magic-string-ast'

/** @deprecated use `generateTransform` instead */
export const getTransformResult = generateTransform

export interface BaseOptions {
  include?: FilterPattern
  exclude?: FilterPattern
  version?: number
}

export function createFilter(options: BaseOptions) {
  return createRollupFilter(options.include, options.exclude)
}

export { normalizePath } from '@rollup/pluginutils'
