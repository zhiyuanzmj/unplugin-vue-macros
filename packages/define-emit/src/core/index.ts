import {
  DEFINE_EMIT,
  HELPER_PREFIX,
  MagicString,
  escapeKey,
  generateTransform,
  isCallOf,
  parseSFC,
  walkASTSetup,
} from '@vue-macros/common'
import type * as t from '@babel/types'

export const EMIT_VARIABLE_NAME = `${HELPER_PREFIX}emit`

export interface Emit {
  name: string
  validator?: string
}

export async function transformDefineEmit(code: string, id: string) {
  if (!code.includes(DEFINE_EMIT)) return

  const { scriptSetup, getSetupAst } = parseSFC(code, id)
  if (!scriptSetup) return

  const offset = scriptSetup.loc.start.offset
  const s = new MagicString(code)
  const setupAst = getSetupAst()!
  const emits: Emit[] = []

  await walkASTSetup(setupAst, (setup) => {
    setup.onEnter(
      (node): node is t.CallExpression => isCallOf(node, DEFINE_EMIT),
      (node, parent) => {
        const [name, validator] = node.arguments
        let emitName: string
        if (!name) {
          if (
            parent?.type !== 'VariableDeclarator' ||
            parent.id.type !== 'Identifier'
          ) {
            throw new Error(
              `A variable must be used to receive the return value of ${DEFINE_EMIT}.`
            )
          }
          emitName = parent.id.name
        } else if (name.type !== 'StringLiteral') {
          throw new Error(
            `The first argument of ${DEFINE_EMIT} must be a string literal.`
          )
        } else {
          emitName = name.value
        }

        emits.push({
          name: emitName,
          validator: validator ? s.sliceNode(validator, { offset }) : undefined,
        })

        s.overwriteNode(
          node,
          `(...args) => ${EMIT_VARIABLE_NAME}(${JSON.stringify(
            emitName
          )}, ...args)`,
          { offset }
        )
      }
    )
  })

  if (emits.length > 0) {
    s.prependLeft(
      offset!,
      `\nconst ${EMIT_VARIABLE_NAME} = defineEmits(${mountEmits()})\n`
    )
  }

  return generateTransform(s, id)

  function mountEmits() {
    const isAllWithoutOptions = emits.every(({ validator }) => !validator)

    if (isAllWithoutOptions) {
      return `[${emits.map(({ name }) => JSON.stringify(name)).join(', ')}]`
    }

    return `{
      ${emits
        .map(
          ({ name, validator }) => `${escapeKey(name)}: ${validator || `null`}`
        )
        .join(',\n  ')}
    }`
  }
}
