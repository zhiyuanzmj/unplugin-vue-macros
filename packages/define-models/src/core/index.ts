import { extractIdentifiers, walkAST } from 'ast-walker-scope'
import {
  DEFINE_EMITS,
  DEFINE_MODELS,
  DEFINE_MODELS_DOLLAR,
  DEFINE_OPTIONS,
  DEFINE_PROPS,
  HELPER_PREFIX,
  MagicString,
  REPO_ISSUE_URL,
  WITH_DEFAULTS,
  generateTransform,
  importHelperFn,
  isCallOf,
  parseSFC,
  resolveObjectKey,
} from '@vue-macros/common'
import {
  type Identifier,
  type LVal,
  type Node,
  type ObjectExpression,
  type ObjectPattern,
  type ObjectProperty,
  type Statement,
  type TSInterfaceBody,
  type TSTypeLiteral,
  type VariableDeclaration,
} from '@babel/types'
import { emitHelperId, useVmodelHelperId } from './helper'

export function transformDefineModels(
  code: string,
  id: string,
  version: number,
  unified: boolean
) {
  let hasDefineProps = false
  let hasDefineEmits = false
  let hasDefineModels = false

  let propsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let propsDestructureDecl: Node | undefined
  let emitsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let emitsIdentifier: string | undefined

  let runtimeDefineFn: string | undefined

  let modelDecl: Node | undefined
  let modelDeclKind: string | undefined
  let modelTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let modelIdentifier: string | undefined
  let modelDestructureDecl: ObjectPattern | undefined

  const modelIdentifiers = new Set<Identifier>()
  const modelVue2: { event: string; prop: string } = { prop: '', event: '' }
  let mode: 'reactivity-transform' | 'runtime' | undefined

  function processDefinePropsOrEmits(node: Node, declId?: LVal) {
    if (isCallOf(node, WITH_DEFAULTS)) {
      node = node.arguments[0]
    }

    let type: 'props' | 'emits'
    if (isCallOf(node, DEFINE_PROPS)) {
      type = 'props'
    } else if (isCallOf(node, DEFINE_EMITS)) {
      type = 'emits'
    } else {
      return false
    }
    const fnName = type === 'props' ? DEFINE_PROPS : DEFINE_EMITS

    if (node.arguments[0]) {
      runtimeDefineFn = fnName
      return false
    }

    if (type === 'props') hasDefineProps = true
    else hasDefineEmits = true

    const typeDeclRaw = node.typeParameters?.params?.[0]
    if (!typeDeclRaw)
      throw new SyntaxError(
        `${fnName}() expected a type parameter when used with ${DEFINE_MODELS}.`
      )

    const typeDecl = resolveQualifiedType(
      typeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!typeDecl) {
      throw new SyntaxError(
        `type argument passed to ${fnName}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    if (type === 'props') propsTypeDecl = typeDecl
    else emitsTypeDecl = typeDecl

    if (declId) {
      if (type === 'props' && declId.type === 'ObjectPattern') {
        propsDestructureDecl = declId
      } else if (type === 'emits' && declId.type === 'Identifier') {
        emitsIdentifier = declId.name
      }
    } else if (type === 'emits') {
      emitsIdentifier = `_${DEFINE_MODELS}_emit`
      s.prependRight(setupOffset + node.start!, `const ${emitsIdentifier} = `)
    }

    return true
  }

  function processDefineModels(
    node: Node,
    declId?: LVal,
    kind?: VariableDeclaration['kind']
  ) {
    if (isCallOf(node, DEFINE_MODELS)) mode = 'runtime'
    else if (isCallOf(node, DEFINE_MODELS_DOLLAR)) mode = 'reactivity-transform'
    else return false

    if (hasDefineModels) {
      throw new SyntaxError(`duplicate ${DEFINE_MODELS}() call`)
    }
    hasDefineModels = true
    modelDecl = node

    const propsTypeDeclRaw = node.typeParameters?.params[0]
    if (!propsTypeDeclRaw) {
      throw new SyntaxError(`expected a type parameter for ${DEFINE_MODELS}.`)
    }
    modelTypeDecl = resolveQualifiedType(
      propsTypeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!modelTypeDecl) {
      throw new SyntaxError(
        `type argument passed to ${DEFINE_MODELS}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    if (mode === 'reactivity-transform' && declId) {
      const ids = extractIdentifiers(declId)
      ids.forEach((id) => modelIdentifiers.add(id))

      if (declId.type === 'ObjectPattern') {
        modelDestructureDecl = declId
        for (const property of declId.properties) {
          if (property.type === 'RestElement') {
            throw new SyntaxError('rest element is not supported')
          }
        }
      } else {
        modelIdentifier = scriptSetup!.loc.source.slice(
          declId.start!,
          declId.end!
        )
      }
    }
    if (kind) modelDeclKind = kind

    return true
  }

  function processDefineOptions(node: Node) {
    if (!isCallOf(node, DEFINE_OPTIONS)) return false

    const [arg] = node.arguments
    if (arg) processVue2Model(arg)

    return true
  }

  function processVue2Script() {
    if (!script) return
    const scriptAst = getScriptAst()!.body
    if (scriptAst.length === 0) return

    // process normal <script>
    for (const node of scriptAst as Statement[]) {
      if (node.type === 'ExportDefaultDeclaration') {
        const { declaration } = node
        if (declaration.type === 'ObjectExpression') {
          processVue2Model(declaration)
        } else if (
          declaration.type === 'CallExpression' &&
          declaration.callee.type === 'Identifier' &&
          ['defineComponent', 'DO_defineComponent'].includes(
            declaration.callee.name
          )
        ) {
          declaration.arguments.forEach((arg) => {
            if (arg.type === 'ObjectExpression') {
              processVue2Model(arg)
            }
          })
        }
      }
    }
  }

  function processVue2Model(node: Node) {
    // model: {
    //   prop: 'checked',
    //   event: 'change'
    // }
    if (node.type !== 'ObjectExpression') return false

    const model = node.properties.find(
      (prop) =>
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        prop.key.name === 'model' &&
        prop.value.type === 'ObjectExpression' &&
        prop.value.properties.length === 2
    ) as ObjectProperty

    if (!model) return false
    ;(model.value as ObjectExpression).properties.forEach((propertyItem) => {
      if (
        propertyItem.type === 'ObjectProperty' &&
        propertyItem.key.type === 'Identifier' &&
        propertyItem.value.type === 'StringLiteral' &&
        ['prop', 'event'].includes(propertyItem.key.name)
      ) {
        const key = propertyItem.key.name as 'prop' | 'event'
        modelVue2[key] = propertyItem.value.value
      }
    })
    return true
  }

  function resolveQualifiedType(
    node: Node,
    qualifier: (node: Node) => boolean
  ) {
    if (qualifier(node)) {
      return node
    }
    if (
      node.type === 'TSTypeReference' &&
      node.typeName.type === 'Identifier'
    ) {
      const refName = node.typeName.name
      const isQualifiedType = (node: Node): Node | undefined => {
        if (
          node.type === 'TSInterfaceDeclaration' &&
          node.id.name === refName
        ) {
          return node.body
        } else if (
          node.type === 'TSTypeAliasDeclaration' &&
          node.id.name === refName &&
          qualifier(node.typeAnnotation)
        ) {
          return node.typeAnnotation
        } else if (node.type === 'ExportNamedDeclaration' && node.declaration) {
          return isQualifiedType(node.declaration)
        }
      }
      for (const node of setupAst) {
        const qualified = isQualifiedType(node)
        if (qualified) {
          return qualified
        }
      }
    }
  }

  function extractPropsDefinitions(node: TSTypeLiteral | TSInterfaceBody) {
    const members = node.type === 'TSTypeLiteral' ? node.members : node.body
    const map: Record<
      string,
      {
        typeAnnotation: string
        options?: Record<string, string>
      }
    > = Object.create(null)

    for (const m of members) {
      if (
        (m.type === 'TSPropertySignature' || m.type === 'TSMethodSignature') &&
        m.key.type === 'Identifier'
      ) {
        const type = m.typeAnnotation?.typeAnnotation
        let typeAnnotation = ''
        let options: Record<string, string> | undefined
        if (type) {
          typeAnnotation += `${m.optional ? '?' : ''}: `
          if (
            type.type === 'TSTypeReference' &&
            type.typeName.type === 'Identifier' &&
            type.typeName.name === 'ModelOptions' &&
            type.typeParameters?.type === 'TSTypeParameterInstantiation' &&
            type.typeParameters.params[0]
          ) {
            typeAnnotation += setupContent.slice(
              type.typeParameters.params[0].start!,
              type.typeParameters.params[0].end!
            )
            if (type.typeParameters.params[1]?.type === 'TSTypeLiteral') {
              options = Object.create(null)
              for (const m of type.typeParameters.params[1].members) {
                if (
                  (m.type === 'TSPropertySignature' ||
                    m.type === 'TSMethodSignature') &&
                  m.key.type === 'Identifier'
                ) {
                  const type = m.typeAnnotation?.typeAnnotation
                  if (type)
                    options![setupContent.slice(m.key.start!, m.key.end!)] =
                      setupContent.slice(type.start!, type.end!)
                }
              }
            }
          } else
            typeAnnotation += `${setupContent.slice(type.start!, type.end!)}`
        }

        map[m.key.name] = { typeAnnotation, options }
      }
    }
    return map
  }

  function getPropKey(key: string, omitDefault = false) {
    if (unified && version === 2 && key === 'modelValue') {
      return 'value'
    }
    return !omitDefault ? key : undefined
  }

  function getEventKey(key: string, omitDefault = false) {
    if (version === 2) {
      if (modelVue2.prop === key) {
        return modelVue2.event
      } else if (key === 'value' || (unified && key === 'modelValue')) {
        return 'input'
      }
    }
    return !omitDefault ? `update:${key}` : undefined
  }

  function rewriteMacros() {
    rewriteDefines()
    if (mode === 'runtime') {
      rewriteRuntime()
    }

    function rewriteDefines() {
      const propsText = Object.entries(map)
        .map(
          ([key, { typeAnnotation }]) => `${getPropKey(key)}${typeAnnotation}`
        )
        .join('\n')

      const emitsText = Object.entries(map)
        .map(
          ([key, { typeAnnotation }]) =>
            `(evt: '${getEventKey(key)}', value${typeAnnotation}): void`
        )
        .join('\n')

      if (hasDefineProps) {
        s.appendLeft(setupOffset + propsTypeDecl!.start! + 1, `${propsText}\n`)
        if (
          mode === 'reactivity-transform' &&
          propsDestructureDecl &&
          modelDestructureDecl
        )
          for (const property of modelDestructureDecl.properties) {
            const text = code.slice(
              setupOffset + property.start!,
              setupOffset + property.end!
            )
            s.appendLeft(
              setupOffset + propsDestructureDecl.start! + 1,
              `${text}, `
            )
          }
      } else {
        let text = ''
        const kind = modelDeclKind || 'let'
        if (mode === 'reactivity-transform') {
          if (modelIdentifier) {
            text = modelIdentifier
          } else if (modelDestructureDecl) {
            text = code.slice(
              setupOffset + modelDestructureDecl.start!,
              setupOffset + modelDestructureDecl.end!
            )
          }
        }

        s.appendRight(
          setupOffset,
          `\n${text ? `${kind} ${text} = ` : ''}defineProps<{
    ${propsText}
  }>();`
        )
      }

      if (hasDefineEmits) {
        s.appendRight(setupOffset + emitsTypeDecl!.start! + 1, `${emitsText}\n`)
      } else {
        emitsIdentifier = `${HELPER_PREFIX}emit`
        s.appendRight(
          setupOffset,
          `\n${
            mode === 'reactivity-transform' ? `const ${emitsIdentifier} = ` : ''
          }defineEmits<{
    ${emitsText}
  }>();`
        )
      }
    }

    function rewriteRuntime() {
      const text = `${importHelperFn(
        s,
        setupOffset,
        'useVModel',
        useVmodelHelperId,
        true
      )}(${Object.entries(map)
        .map(([name, { options }]) => {
          const prop = getPropKey(name, true)
          const evt = getEventKey(name, true)
          if (!prop && !evt && !options) return stringifyValue(name)

          const args = [name, prop, evt].map((arg) => stringifyValue(arg))
          if (options) {
            const str = Object.entries(options)
              .map(([k, v]) => `  ${stringifyValue(k)}: ${v}`)
              .join(',\n')
            args.push(`{\n${str}\n}`)
          }

          return `[${args.join(', ')}]`
        })
        .join(', ')})`
      s.overwriteNode(modelDecl!, text, { offset: setupOffset })
    }
  }

  function processAssignModelVariable() {
    if (!emitsIdentifier)
      throw new Error(
        `Identifier of returning value of ${DEFINE_EMITS} is not found, please report this issue.\n${REPO_ISSUE_URL}`
      )

    function overwrite(
      node: Node,
      id: Identifier,
      value: string,
      original = false
    ) {
      const eventName = aliasMap[id.name]
      const content = `${importHelperFn(
        s,
        setupOffset,
        'emitHelper',
        emitHelperId,
        true
      )}(${emitsIdentifier}, '${getEventKey(String(eventName))}', ${value}${
        original ? `, ${id.name}` : ''
      })`
      s.overwriteNode(node, content, { offset: setupOffset })
    }

    walkAST(setupAst!, {
      leave(node) {
        if (node.type === 'AssignmentExpression') {
          if (node.left.type !== 'Identifier') return
          const id = this.scope[node.left.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          const left = s.sliceNode(node.left, { offset: setupOffset })
          let right = s.sliceNode(node.right, { offset: setupOffset })
          if (node.operator !== '=') {
            right = `${left} ${node.operator.replace(/=$/, '')} ${right}`
          }

          overwrite(node, id, right)
        } else if (node.type === 'UpdateExpression') {
          if (node.argument.type !== 'Identifier') return
          const id = this.scope[node.argument.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          let value = node.argument.name
          if (node.operator === '++') value += ' + 1'
          else value += ' - 1'

          overwrite(node, id, value, !node.prefix)
        }
      },
    })
  }

  if (!code.includes(DEFINE_MODELS)) return
  const { script, scriptSetup, getSetupAst, getScriptAst } = parseSFC(code, id)
  if (!scriptSetup) return

  const setupOffset = scriptSetup.loc.start.offset
  const setupContent = scriptSetup.content
  const setupAst = getSetupAst()!.body

  const s = new MagicString(code)

  if (version === 2) processVue2Script()

  // process <script setup>
  for (const node of setupAst) {
    if (node.type === 'ExpressionStatement') {
      processDefinePropsOrEmits(node.expression)

      if (version === 2) {
        processDefineOptions(node.expression)
      }

      if (
        processDefineModels(node.expression) &&
        mode === 'reactivity-transform'
      )
        s.remove(node.start! + setupOffset, node.end! + setupOffset)
    } else if (node.type === 'VariableDeclaration' && !node.declare) {
      const total = node.declarations.length
      let left = total

      for (let i = 0; i < total; i++) {
        const decl = node.declarations[i]
        if (decl.init) {
          processDefinePropsOrEmits(decl.init, decl.id)

          if (
            processDefineModels(decl.init, decl.id, node.kind) &&
            mode === 'reactivity-transform'
          ) {
            if (left === 1) {
              s.remove(node.start! + setupOffset, node.end! + setupOffset)
            } else {
              let start = decl.start! + setupOffset
              let end = decl.end! + setupOffset
              if (i < total - 1) {
                // not the last one, locate the start of the next
                end = node.declarations[i + 1].start! + setupOffset
              } else {
                // last one, locate the end of the prev
                start = node.declarations[i - 1].end! + setupOffset
              }
              s.remove(start, end)
              left--
            }
          }
        }
      }
    }
  }

  if (!modelTypeDecl) return

  if (runtimeDefineFn)
    throw new SyntaxError(
      `${runtimeDefineFn}() cannot accept non-type arguments when used with ${DEFINE_MODELS}()`
    )

  if (modelTypeDecl.type !== 'TSTypeLiteral') {
    throw new SyntaxError(
      `type argument passed to ${DEFINE_MODELS}() must be a literal type, or a reference to an interface or literal type.`
    )
  }

  const map = extractPropsDefinitions(modelTypeDecl)
  const aliasMap: Record<string, string | number> = Object.create(null)
  if (modelDestructureDecl)
    for (const p of modelDestructureDecl.properties) {
      if (p.type !== 'ObjectProperty') continue
      try {
        const key = resolveObjectKey(p)
        if (p.value.type !== 'Identifier') continue
        aliasMap[p.value.name] = key
      } catch {}
    }

  // const defaults = resolveObjectExpression(defaultsAst)
  // if (!defaults) return { defaultsAst }
  rewriteMacros()

  if (mode === 'reactivity-transform' && hasDefineModels)
    processAssignModelVariable()

  return generateTransform(s, id)
}

function stringifyValue(value: string | undefined) {
  return value !== undefined ? JSON.stringify(value) : 'undefined'
}
