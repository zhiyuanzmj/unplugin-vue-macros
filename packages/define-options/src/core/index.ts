import {
  DEFINE_OPTIONS,
  MagicString,
  addNormalScript,
  checkInvalidScopeReference,
  generateTransform,
  importHelperFn,
  parseSFC,
} from '@vue-macros/common'
import { walkAST } from 'ast-walker-scope'
import {
  type ExportDefaultDeclaration,
  type Program,
  type Statement,
} from '@babel/types'
import { filterMacro, hasPropsOrEmits as hasDisallowedProp } from './utils'

export * from './utils'

export function transformDefineOptions(code: string, id: string) {
  if (!code.includes(DEFINE_OPTIONS)) return

  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return
  const { scriptSetup, getSetupAst, getScriptAst } = sfc
  const setupOffset = scriptSetup.loc.start.offset
  const setupAst = getSetupAst()!

  const nodes = filterMacro(setupAst!.body)
  if (nodes.length === 0) {
    return
  } else if (nodes.length > 1)
    throw new SyntaxError(`duplicate ${DEFINE_OPTIONS}() call`)

  const scriptAst = getScriptAst()!
  if (scriptAst) checkDefaultExport(scriptAst.body)

  const setupBindings = getRootScopeIds(setupAst)

  const s = new MagicString(code)

  const [node] = nodes
  const [arg] = node.arguments
  if (arg) {
    const normalScript = addNormalScript(sfc, s)

    const scriptOffset = normalScript.start()

    s.appendLeft(
      scriptOffset,
      `\nexport default /*#__PURE__*/ ${importHelperFn(
        s,
        scriptOffset,
        'defineComponent'
      )}(`
    )

    if (arg.type === 'ObjectExpression' && hasDisallowedProp(arg))
      throw new SyntaxError(
        `${DEFINE_OPTIONS}() please use defineProps, defineEmits, defineExpose, or defineSlots instead.`
      )

    checkInvalidScopeReference(arg, DEFINE_OPTIONS, setupBindings)

    s.moveNode(arg, scriptOffset, { offset: setupOffset })

    // removes defineOptions()
    s.remove(setupOffset + node.start!, setupOffset + arg.start!)
    s.remove(setupOffset + arg.end!, setupOffset + node.end!)

    s.appendRight(scriptOffset, ');')
    normalScript.end()
  } else {
    // removes defineOptions()
    s.removeNode(node, { offset: setupOffset })
  }

  return generateTransform(s, id)
}

function checkDefaultExport(stmts: Statement[]) {
  const hasDefaultExport = stmts.some(
    (node): node is ExportDefaultDeclaration =>
      node.type === 'ExportDefaultDeclaration'
  )
  if (hasDefaultExport)
    throw new SyntaxError(
      `${DEFINE_OPTIONS} cannot be used with default export within <script>.`
    )
}

function getRootScopeIds(program: Program) {
  let ids: string[] = []
  walkAST(program, {
    enter(node) {
      if (node.type === 'BlockStatement') {
        this.skip()
      }
    },
    leave(node) {
      if (node.type !== 'Program') return
      ids = Object.keys(this.scope)
    },
  })

  return ids
}
