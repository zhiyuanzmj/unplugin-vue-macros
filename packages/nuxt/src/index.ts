import { defineNuxtModule, useNuxt } from '@nuxt/kit'
import VueMacros from 'unplugin-vue-macros/vite'
import {
  type Options as OptionsShortVmodel,
  transformShortVmodel,
} from '@vue-macros/short-vmodel'
import { transformBooleanProp } from '@vue-macros/boolean-prop'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import type {} from '@nuxt/devtools'
import { type VolarOptions } from '@vue-macros/volar'

export type VueMacrosOptions = Options & {
  booleanProp?: {} | false
  shortVmodel?: OptionsShortVmodel | false
}

export default defineNuxtModule<VueMacrosOptions>({
  meta: {
    name: 'unplugin-vue-macros',
    configKey: 'macros',
  },
  defaults: {},
  setup(options) {
    const nuxt = useNuxt()
    const resolvedOptions = resolveOptions(options)

    nuxt.hook('vite:extendConfig', (config, { isClient }) => {
      function findPluginAndRemove(name: string): Plugin | undefined {
        const idx = config.plugins!.findIndex(
          (plugin) => plugin && 'name' in plugin && plugin.name === name
        )
        if (idx === -1) return
        const plugin = config.plugins![idx]
        config.plugins!.splice(idx, 1)
        return plugin as any
      }
      config.plugins ||= []
      const vue = findPluginAndRemove('vite:vue')
      const vueJsx = findPluginAndRemove('vite:vue-jsx')

      config.plugins.push(
        VueMacros({
          ...resolvedOptions,
          plugins: { vue, vueJsx },
          nuxtContext: { isClient },
        })
      )
    })

    nuxt.hook('prepare:types', (opts) => {
      opts.references.push({ types: 'unplugin-vue-macros/macros-global' })
    })

    nuxt.hook('devtools:customTabs', (tabs) => {
      tabs.push({
        name: 'vue-macros',
        title: 'Vue Macros',
        icon: 'https://raw.githubusercontent.com/vue-macros/vue-macros/main/docs/public/favicon.svg',
        view: {
          type: 'iframe',
          src: '/__vue-macros',
        },
      })
    })

    nuxt.options.typescript.tsConfig ||= {}

    // @ts-expect-error https://github.com/unjs/pkg-types/pull/130
    nuxt.options.typescript.tsConfig.vueCompilerOptions ||= {}
    const vueCompilerOptions =
      // @ts-expect-error
      nuxt.options.typescript.tsConfig.vueCompilerOptions

    vueCompilerOptions.vueMacros ||= {}
    const volarOptions = vueCompilerOptions.vueMacros as VolarOptions

    vueCompilerOptions.plugins ||= []
    const volarPlugins = vueCompilerOptions.plugins

    if (resolvedOptions.defineOptions)
      volarPlugins.push('@vue-macros/volar/define-options')

    if (resolvedOptions.defineSlots)
      volarPlugins.push('@vue-macros/volar/define-slots')

    if (resolvedOptions.defineModels)
      volarPlugins.push('@vue-macros/volar/define-models')

    if (resolvedOptions.defineProps)
      volarPlugins.push('@vue-macros/volar/define-props')

    if (resolvedOptions.definePropsRefs)
      volarPlugins.push('@vue-macros/volar/define-props-refs')

    if (resolvedOptions.exportProps)
      volarPlugins.push('@vue-macros/volar/export-props')

    if (resolvedOptions.jsxDirective)
      volarPlugins.push('@vue-macros/volar/jsx-directive')

    if (resolvedOptions.defineProp)
      vueCompilerOptions.experimentalDefinePropProposal =
        resolvedOptions.defineProp.edition || 'kevinEdition'

    nuxt.options.vite.vue ||= {}
    nuxt.options.vite.vue.include ||= [/\.vue$/]
    if (!Array.isArray(nuxt.options.vite.vue.include))
      nuxt.options.vite.vue.include = [nuxt.options.vite.vue.include]
    nuxt.options.vite.vue.include.push(/\.setup\.[cm]?[jt]sx?$/)

    if (options.shortVmodel !== false || options.booleanProp !== false) {
      nuxt.options.vite.vue.template ||= {}
      nuxt.options.vite.vue.template.compilerOptions ||= {}
      nuxt.options.vite.vue.template.compilerOptions.nodeTransforms ||= []
      const { nodeTransforms } = nuxt.options.vite.vue.template.compilerOptions

      if (options.shortVmodel !== false) {
        volarPlugins.push('@vue-macros/volar/short-vmodel')
        nodeTransforms.push(transformShortVmodel(options.shortVmodel))
        if (options.shortVmodel) {
          volarOptions.shortVmodel = {
            prefix: options.shortVmodel.prefix,
          }
        }
      }

      if (options.booleanProp !== false)
        nodeTransforms.push(transformBooleanProp())
    }
  },
})

declare module '@nuxt/schema' {
  interface NuxtConfig {
    macros?: Options
  }
  interface NuxtOptions {
    macros?: Options
  }
}
