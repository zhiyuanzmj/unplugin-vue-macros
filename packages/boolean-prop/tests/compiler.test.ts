import { compileTemplate } from '@vue/compiler-sfc'
import { describe, expect, test } from 'vitest'
import { transformBooleanProp } from '../src'

function compile(code: string) {
  return compileTemplate({
    source: code,
    filename: 'anonymous.vue',
    id: 'xxx',
    compilerOptions: {
      nodeTransforms: [transformBooleanProp()],
    },
  })
}

describe('compiler', () => {
  test('basic', () => {
    const original = compile(`<input :checked="true" />`)
    const sugar = compile(`<input check />`)
    expect(sugar.code).toBe(original.code)
    expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
  })
})
