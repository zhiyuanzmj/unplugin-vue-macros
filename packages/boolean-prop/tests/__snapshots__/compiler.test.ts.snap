// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler > basic 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "checked",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": "check",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 3,
    "content": "true",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": "check",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "source": "check",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "bind",
  "type": 7,
}
`;
